### ISSUE TITLE
JSDoc @implements clause not checked

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#30156]
```js
// @ts-check

/**
 * @typedef Foo
 * @property foo {() => string}
 */

/**
 * @implements {Foo}
 */
class C1 {
    foo = () => {
        return 10;
    }
}

/**
 * @implements {Foo}
 */
class C2 {
    foo() {
        return 10;
    }
}
```

**Expected**: At least an error on C1
**Actual**: Neither of these error

**Comment by brendankenny** [comment]
also see #17498, though it doesn't have the clearest initial comment describing the problem :)

**Comment by TimvdLippe** [comment]
FYI this does work with an `@interface` now, since #35629 has been implemented.

**Comment by Paril** [comment]
I attempted this with a TS `interface` but was unable to get it working. ie:
```ts
interface Test
{
func(n: number): void
}
```

```js
/** @implements {Test} */
class MyTest
{
func(n) // `n` is implicit any
{
}
}
```

**Comment by kevinfiol** [comment]
> 
> 
> I attempted this with a TS `interface` but was unable to get it working. ie:
> 
> ```ts
> interface Test
> {
> func(n: number): void
> }
> ```
> 
> ```js
> /** @implements {Test} */
> class MyTest
> {
> func(n) // `n` is implicit any
> {
> }
> }
> ```
> 
> EDIT: okay, so it technically _does_ work (in that MyTest can be passed to things expecting a Test, and as long as you don't re-declare it you will see the inherited type), but "func" doesn't auto-inherit the documentation or type annotations. I think that's a different issue though? maybe?

Piggybacking on this. I thought I would get my answer in this thread, but also having this issue. Tried `@implements` and `@interface` but no dice. The best workaround I found was [this](https://github.com/microsoft/TypeScript/issues/17498#issuecomment-399439654) comment where you set the type of the `this` reference to the interface, but this is a very imperfect solution if you're looking to get good IDE integration since the class methods don't pick up their types from the interface.

**Comment by Dd-ouni** [comment]
> 解决方法，基于[此注释](https://github.com/microsoft/TypeScript/issues/17498#issuecomment-399439654)。
> 
> ```js
> class MyStoreImpl {
>   constructor() {
>     // Verify that the class implements the interface.
>     void (/** @type {import('./types').MyStore} */ (this));
>   }
> }
> ```

Hello, have you solved this problem
```
    constructor(){
    // Verify that the class implements the interface.
    void (/** @type {import('webpack').LoaderOptionsPlugin} */ (this));
    }
    
}
```
vscode IntelliSense not working

**Comment by MSFTKonnor** [comment]
+1. Still hitting this problem in TypeScript 4.9.2 as well as TypeScript 5.0.

**Comment by david-sanford-deel** [comment]
+1.

**Comment by RyanCavanaugh** [comment]
Errors now

```
src/a.js:12:5 - error TS2416: Property 'foo' in type 'C1' is not assignable to the same property in base type 'Foo'.
  Type '() => number' is not assignable to type '() => string'.
    Type 'number' is not assignable to type 'string'.

12     foo = () => {
       ~~~

src/a.js:21:5 - error TS2416: Property 'foo' in type 'C2' is not assignable to the same property in base type 'Foo'.
  Type '() => number' is not assignable to type '() => string'.
    Type 'number' is not assignable to type 'string'.

21     foo() {
       ~~~

```

**Comment by henrahmagix** [comment]
Amazing, thanks @RyanCavanaugh!
